perf: improve import

letsfindaway opened the pull request:
This is a draft PR for testing performance improvements on import:

- do not create thumbnail files during import
- thumbnail files are created later when populating the thumbnails
- do not save the TOC after each imported page
- immediately use image after rendering thumbnail instead reading it from the file afterwards
- only create a missing thumbnail when needed
- avoid repeated check for existence of PDF file when saving scene
- do not check for existence of PDF file when writing SVG
- assume file always exists
- there is no way to create a PDF item without an existing file
- use media assets instead of hardcoded directory strings
- fix: use media asset uuid when importing
- create media asset uuid in UBPersistenceManager::addFileToDocument
- return uuid on success

On my system it improves the time to import a 2000 pages PDF document from 45 sec to 7 sec.

@kaamui Please test this on your network storage environment. But as this PR is mainly about reducing I/O, it should also give a great performance boost there. It also reduces CPU usage as the imported pages are only rendered after import.

letsfindaway commented:
If we could find a way to be sure that the PDF document already exists in `UBSvgSubsetWriter::pdfItemToLinkedPDF` then we could even omit the check there and save another `QFile::exists`.

kaamui commented:
tested with the patch before the two last commits : 

<img width="1097" height="637" alt="image" src="https://github.com/user-attachments/assets/e5bcfb35-67c3-4526-87de-eb402eba89ff" />

letsfindaway commented:
> Similar result with updated patch : <img alt="image" width="1092" height="136" src="https://private-user-images.githubusercontent.com/11720903/479949369-d7937640-504a-4e54-b767-6f16d7e83a15.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU2ODUyODMsIm5iZiI6MTc1NTY4NDk4MywicGF0aCI6Ii8xMTcyMDkwMy80Nzk5NDkzNjktZDc5Mzc2NDAtNTA0YS00ZTU0LWI3NjctNmYxNmQ3ZTgzYTE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODIwVDEwMTYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjMDJjMzg3NGU4YjE2NjZjYWU1ZTlhNzQ5ZDhhMWU4MWU0OWExYmEwZTA4NmRmZjE2ZGRlMzJhYjYzNjdhZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.olGqNTLW2jQLZhtFaiwTy5nzU7JdcdOdW8Ol2DN6mVU">

yes, this is expected. The two later patches do not add as much as the first one.

letsfindaway commented:
With my latest commit things improved dramatically. On my system the import now takes incredible 540ms for the 2000 pages document!! I also added a small delay between scans of pages. This improved user experience very much and only increases scanning time by about 10%.

What I did:

- Scenes created during import or loaded during scanning are no longer cached.
- Use such uncached scenes to generate thumbnails.
- The TOC was saved after each page. This now only occurs at the end.
- Generate thumbnails using the asset scanner, as we here already load uncached scenes.
- To do this, also use the TOC "has no assets entry" to optionally create a thumbnail if it does not exist.
- After scanning a page, I introduced a 5ms delay to give the main thread time processing user interactions.
- Make sure that the `UBDocument` instance used during import remains alive afterwards.
- Allow to create a `UBThumbnailScene` in `UBDocument` without immediately starting the thumbnail loader.

letsfindaway commented:
> Issue with thumbnails in Documents Mode (WIP)

fixed with last commit

letsfindaway commented:
With some more time, I now removed the last commit, which was breaking many things, and replaced it with better code. Still have to do more tests, but the main things seem to be much better now.

letsfindaway commented:
Added some small fixes, rearranged and squashed commits.

I think this is now a sound base for further tests.

letsfindaway commented:
The latest commit should again boost performance, especially for thumbnail loading. I observed that on a fast system the decoding of the thumbnail JPG took more time than everything else in the thumbnail loading process. As this is just CPU usage I now moved the decoding to the background loader, which now returns a `QVariant` with either the plain data (for scene loading) or a decoded pixmap. This relieves the main thread to such an extent that no special delay is necessary after loading a thumbnail.

I also fine-tuned asset scanning. This now starts already in Document mode when selecting a document, but is aborted when you select another document. However the intermediate results are kept. Clean-up of the asset scanner in this case was fixed.

kaamui commented on src/frameworks/UBBlockingBuffer.h:
yep, mine complained ^^

kaamui commented:
Just made some quick tests, sorry I don't have the time to fully qualify them : 
- Sometimes, clicking on import or export during import scanning will block until scanning is done, and then open the system dialog
- systematic crash on `UBGraphicsScene::loadingCompleted` (called by `UBSceneCache::SceneCacheEntry::startLoading()`) because `this` is null, during import
- same crash after relaunching OpenBoard, go to the "broken document" and double-clicking on one of the not already loaded page
- replacement dialog visually changed, but its height seems a bit odd.

<img width="1463" height="865" alt="image" src="https://github.com/user-attachments/assets/4839f146-95a9-475f-923a-319aa3098cb1" />


I find the targeted behavior really great. The fact that we can import a file, partially generate the thumbnails, exit the application, relaunch it and resume thumbnail generation from where it sopped is definitely a good thing.

letsfindaway commented:
I just added an experimental commit adding a function to wait for unsaved data in the persistence worker. Please check whether this fixes the crash.

If yes, then I will anyway add a guard against `nullptr` for the scene, but before that I would like to know whether this commit helps.

I already could reproduce the crash by providing a document with a missing or empty SVG file, which was however listed in the TOC. What should we do in this case? Silently delete this page from the TOC, as its content cannot be reproduced? Or create an empty page instead?

letsfindaway commented:
There is another corner case with the persistence worker: what if a scene is in the queue of the persistence worker, but before it is written, we either have another immediate, unqueued save of the same scene or we even delete the scene? Then it is later overwritten when the worker saves it. It could even happen that a non-queued saving action and saving by the worker occur at the same time!

I think we must make sure that a scene is not currently saved or waiting to be saved by the persistence worker, before either unqueued saving or deleting a scene.

Note that this corner case is not new with TOC, but already existed before. It might be unlikely, but it can happen. I will add another function to the `UBPersistenceWorker` which guarantees that a specified scene is not saving or pending saving.

letsfindaway commented:
The latest commit introduces more checks for a nullptr scene. Especially the `UBDocument::sceneLoaded()` isnow no longer called via the `UBGraphicsScene` but directly from the `UBSceneCacheEntry` when loading is either completed successfully or was not possible and ended in an error.

Now also `nullptr` checks are introduced in all functions retrieving a scene.

kaamui commented:
Hi @letsfindaway,

The systematic crash is gone. The overall user experience is better. I still experienced once the scan process blocking import dialog until it ends, but it seems rare and not easily reproducible. 

Also tested on Windows, with Qt 6.9.1, on both local and distant storage, and despite some actions take longer than before in some contexts (createDocumentProxiesStructure at OpenBoard launch takes longer than with 1.7.3 in distant storage on Windows for example), the fact that the user can interact during the scanning process makes the experience way better.

In the worst cases, the lag when writing on Board is still important but non blocking, and again, import of very large documents should not happen in the context of a class so it seems OK to me at this point (import on distant storage takes an average of 3 blocking minutes in 1.7.3, 5 minutes in dev+1311.patch, but user can start using the document within ~45s)

Thank you for this work.
